aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -884,6 +884,9 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
 }
 
 Tensor linalg_cholesky(const Tensor &self) {
+  if (self.numel() == 0) {
+    return at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  }
```

Comment:
Could you please remove this? cuSOLVER calls should handle this case correctly if the lda parameter is right. We don't need to add this early return ourselves.

Reply:
Zero matrix stride or zero batch size cause problem in the below `arange` call. Writing an `if` statement there would make no difference than an `if` statement here.

Besides that, if we can shortcut the functionality at a much earlier place, why do we want to rely on the much deeper implementation of LAPACK, cusolver, or MAGMA with zero batch/matrix size? We don't even have control to those implementations, and they may burn us in the future.

Reply:
Matrix of size 0x0 is not a special case and shouldn't be treated as such, I don't think 0xNxN should be treated as special either.

> why do we want to rely on the much deeper implementation of LAPACK, cusolver, or MAGMA with zero batch/matrix size? We don't even have control to those implementations, and they may burn us in the future.

LAPACK API is de facto standard and I doubt that vendor implementations would deviate from reference LAPACK. Existing calls to LAPACK and MAGMA work without this early return. cuSOLVER's potrf will work as well. It's better to keep this kind of thing closer to the source of the problem and leaving a comment explaining why it's needed to be treated as a special case. In this case, I think it should be only in `apply_cholesky_lib_potrfBatched` and the problem is not in cuSOLVER's potrfBatched, it happily accepts batch size 0 and would error on negative values, but in arange.
Either an early return could be added there or arange call modified to something like
```cpp
  // cusolver batched kernels require input be "device array of device pointers"
  Tensor self_working_copy_array = at::arange(
    reinterpret_cast<int64_t>(self_data),
    reinterpret_cast<int64_t>(&self_data[std::max<int>(batch_size - 1,  0) * matrix_stride]) + 1,
    static_cast<int64_t>(std::max<int64_t>(matrix_stride, 1) * sizeof(scalar_t)),// step should be non-zero
    self_working_copy.options().dtype(at::kLong));
```

Reply:
@mruberry What are your thoughts?

Reply:
Short-circuiting on special cases also seems natural and readable.

Reply:
> potrfBatched says batchSize < 1 is an error https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-batchpotrf

My bad, I mistakenly looked at `potrsBatched` which errors on batchSize < 0. Other functions that allow batchSize = 0 are `gesvdjBatched`, `syevjBatched`, `trsmBatched`, `gelsBatched`, and others from cuBLAS. So I'd say this is a bug for `potrfBatched` that it errors on batchSize < 1 😃 

I still think it's better to keep short-circuiting closer to the source of the problem and it should be put next to the `potrfBatched` call. If someone decides to use these lower-level calls instead of the "front-end" linalg_cholesky function to develop other functionality then it puts the responsibility on that developer to know that this function doesn't work for zero batched case.